controller/registries/nacos/v2config.go

```diff
@@ -0,0 +1,155 @@
+// Copyright The HTNN Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nacos
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/nacos-group/nacos-sdk-go/v2/clients"
+	"github.com/nacos-group/nacos-sdk-go/v2/common/constant"
+	"github.com/nacos-group/nacos-sdk-go/v2/vo"
+
+	registrytype "mosn.io/htnn/types/pkg/registry"
+	"mosn.io/htnn/types/registries/nacos"
+)
+
+type Client struct {
+	Client any
+}
+
+func (reg *Nacos) newV2Client(config *nacos.Config) (*nacosClient, error) {
+	uri, err := url.Parse(config.ServerUrl)
+	if err != nil {
+		return nil, fmt.Errorf("invalid server url: %s", config.ServerUrl)
+	}
+
+	domain := uri.Hostname()
+	p := uri.Port()
+	port := defaultNacosPort
+	if p != "" {
+		port, _ = strconv.Atoi(p)
+	}
+
+	cc := constant.NewClientConfig(
+		constant.WithTimeoutMs(defaultTimeoutMs),
+		constant.WithLogLevel(defaultLogLevel),
+		constant.WithNotLoadCacheAtStart(defaultNotLoadCache),
+		constant.WithLogRollingConfig(&constant.ClientLogRollingConfig{
+			MaxSize:    defaultLogMaxSizeMB,
+			MaxAge:     defaultLogMaxDays,
+			MaxBackups: defaultLogMaxBackups,
+		}),
+		// To simplify the permissions, use the path under current work dir to store log & cache
+	)
+
+	sc := []constant.ServerConfig{
+		*constant.NewServerConfig(domain, uint64(port),
+			constant.WithScheme(uri.Scheme),
+		),
+	}
+
+	namingClient, err := clients.NewNamingClient(vo.NacosClientParam{
+		ClientConfig:  cc,
+		ServerConfigs: sc,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("can not create naming client, err: %v", err)
+	}
+
+	if config.Namespace == "" {
+		config.Namespace = "public"
+	}
+	if len(config.Groups) == 0 {
+		config.Groups = []string{"DEFAULT_GROUP"}
+	}
```

Comment:
The code above is almost copied from the `config.go`. Could we share as much as possible?
For example, we can construct the project like this:
```
nacos/
   client/
     v1/
     v2/
```

the package v1 implements v1 client, and v2 implements v2 client. The client is an interface that provides the methods which are different in v1/v2. The common part will be outside of the client interface.

Reply:
the use of nacos api is different.
v1 use "github.com/nacos-group/nacos-sdk-go"
v2 use "github.com/nacos-group/nacos-sdk-go/v2"

Reply:
@lyt122 
What about writing code like this:
```
func (reg *Nacos) Start(c registrytype.RegistryConfig) error {
	config := c.(*nacos.Config)
	reg.version = config.Version
        // this method returns a client interface
	client, err := reg.newClient(config)
	if err != nil {
		return err
	}
        // call client's fetchAllServices inside
	fetchedServices, err := reg.fetchAllServices(client)
	if err != nil {
		return fmt.Errorf("fetch all services error: %v", err)
	}
	reg.client = client

	for key := range fetchedServices {
                 // // call client's subscribe inside
		err = reg.subscribe(key.GroupName, key.ServiceName)
		if err != nil {
			reg.logger.Errorf("failed to subscribe service, err: %v, service: %v", err, key)
			// the service will be resubscribed after refresh interval
			delete(fetchedServices, key)
		}
	}

	reg.watchingServices = fetchedServices

	dur := 30 * time.Second
	refreshInteval := config.GetServiceRefreshInterval()
	if refreshInteval != nil {
		dur = refreshInteval.AsDuration()
	}
```

Reply:
@spacewander 
The subscribe function need the `getSubscribeCallback` function, which need `store` in Nacos to update serviceEntry.
And other params also is needed, such as  `stopped`, `logger`. 
Should I write like this
```
subscribe(groupName string, serviceName string, stopped atomic.Bool, logger log.RegistryLogger, store registry.ServiceEntryStore)
```
or add these params in Client struct?

Reply:
What about passing the result of getSubscribeCallback? In this way only one param is needed.
We can define our own model.SubscribeService, which is the common part between Nacos v1 & v2's models. Then we can add an adapter to wrap the result of getSubscribeCallback.

Reply:
v1's callback use model.SubscribeService
```
type SubscribeService struct {
	ClusterName string            `json:"clusterName"`
	Enable      bool              `json:"enable"`
	InstanceId  string            `json:"instanceId"`
	Ip          string            `json:"ip"`
	Metadata    map[string]string `json:"metadata"`
	Port        uint64            `json:"port"`
	ServiceName string            `json:"serviceName"`
	Valid       bool              `json:"valid"`
	Weight      float64           `json:"weight"`
	Healthy     bool              `json:"healthy"`
}
```
but v2's callback use model.Instance
```
type Instance struct {
    InstanceId                string            `json:"instanceId"`
    Ip                        string            `json:"ip"`
    Port                      uint64            `json:"port"`
    Weight                    float64           `json:"weight"`
    Healthy                   bool              `json:"healthy"`
    Enable                    bool              `json:"enabled"`
    Ephemeral                 bool              `json:"ephemeral"`
    ClusterName               string            `json:"clusterName"`
    ServiceName               string            `json:"serviceName"`
    Metadata                  map[string]string `json:"metadata"`
    InstanceHeartBeatInterval int               `json:"instanceHeartBeatInterval"`
    IpDeleteTimeout           int               `json:"ipDeleteTimeout"`
    InstanceHeartBeatTimeOut  int               `json:"instanceHeartBeatTimeOut"`
}
```
we can't define the common part

Reply:
We only use Ip, Port and Metadata in getSubscribeCallback. These fields exits both in v1 and v2, so it is possible to define the common part